core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java

```diff
@@ -0,0 +1,80 @@
+package com.sequenceiq.cloudbreak.service;
+
+import com.sequenceiq.cloudbreak.domain.stack.loadbalancer.TargetGroup;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.cmtemplate.CmTemplateProcessor;
+import com.sequenceiq.cloudbreak.cmtemplate.configproviders.knox.KnoxRoles;
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.domain.stack.cluster.Cluster;
+import com.sequenceiq.cloudbreak.domain.stack.instance.InstanceGroup;
+import com.sequenceiq.cloudbreak.template.model.ServiceComponent;
+import com.sequenceiq.common.api.type.InstanceGroupType;
+import com.sequenceiq.common.api.type.LoadBalancerType;
+
+@Service
+public class LoadBalancerConfigService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(LoadBalancerConfigService.class);
+
+    private static final String ENDPOINT_SUFFIX = "gateway";
+
+    private static final String PUBLIC_SUFFIX = "external";
+
+    private static final Set<Integer> DEFAULT_KNOX_PORTS = Set.of(443);
+
+    public Set<String> getKnoxGatewayGroups(Stack stack) {
+        LOGGER.info("Fetching list of instance groups with Knox gateway installed");
+        Set<String> groupNames = new HashSet<>();
+        Cluster cluster = stack.getCluster();
+        if (cluster != null) {
+            CmTemplateProcessor cmTemplateProcessor = new CmTemplateProcessor(cluster.getBlueprint().getBlueprintText());
+            Map<String, Set<ServiceComponent>> components = cmTemplateProcessor.getServiceComponentsByHostGroup();
+            LOGGER.info("Checking if Knox gateway is explicitly defined");
+            groupNames = components.entrySet().stream()
+                .filter(e -> e.getValue().stream().anyMatch(c -> KnoxRoles.KNOX_GATEWAY.equals(c.getComponent())))
```

Comment:
**suggestion:** I'd generally prefer not to double nest streams like this, but I'm not sure it's avoidable here.

You _could_ extract a function `hostGroupContainsKnoxGateway` that contains the `e -> e.getValue().stream().anyMatch(c -> KnoxRoles.KNOX_GATEWAY.equals(c.getComponent()))` portion and then do:
```java
            groupNames = components.entrySet().stream()
                .filter(hostGroupContainsKnoxGateway)
                .map(Map.Entry::getKey)
                .collect(Collectors.toSet());
```

But, barring that, consider expanding the lambda function parameters `e` and `c` into what they they actually represent -- `entry` and `serviceComponent` I think.

Reply:
I can do a separate function to make it more readable.